plutarch-api/src/Plutarch/Api.hs

```diff
@@ -0,0 +1,1275 @@
+  deriving stock
+    ( -- | @since 2.0.0
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since 2.0.0
+      PlutusType
+    , -- | @since 2.0.0
+      PIsData
+    , -- | @since 2.0.0
+      PEq
+    , -- | @since 2.0.0
+      PShow
+    )
+
+-- | @since 2.0.0
+instance DerivePlutusType POutputDatum where
+  type DPTStrat _ = PlutusTypeData
+
+-- | @since 2.0.0
+instance PUnsafeLiftDecl POutputDatum where
+  type PLifted POutputDatum = Plutus.OutputDatum
+
+-- | @since 2.0.0
+deriving via
+  (DerivePConstantViaData Plutus.OutputDatum POutputDatum)
+  instance
+    PConstantDecl Plutus.OutputDatum
+
+-- | @since 2.0.0
+newtype PDatum (s :: S) = PDatum (Term s PData)
+  deriving stock
+    ( -- | @since 2.0.0
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since 2.0.0
+      PlutusType
+    , -- | @since 2.0.0
+      PIsData
+    , -- | @since 2.0.0
+      PEq
+    , -- | @since 2.0.0
+      PShow
+    )
+
+-- | @since 2.0.0
+instance DerivePlutusType PDatum where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since 2.0.0
+instance PUnsafeLiftDecl PDatum where
+  type PLifted PDatum = Plutus.Datum
+
+-- | @since 2.0.0
+deriving via
+  (DerivePConstantViaBuiltin Plutus.Datum PDatum PData)
+  instance
+    PConstantDecl Plutus.Datum
+
+-- | @since 2.0.0
+newtype PRedeemer (s :: S) = PRedeemer (Term s PData)
+  deriving stock
+    ( -- | @since 2.0.0
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since 2.0.0
+      PlutusType
+    , -- | @since 2.0.0
+      PIsData
+    , -- | @since 2.0.0
+      PEq
+    , -- | @since 2.0.0
+      PShow
+    )
+
+-- | @since 2.0.0
+instance DerivePlutusType PRedeemer where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since 2.0.0
+instance PUnsafeLiftDecl PRedeemer where
+  type PLifted PRedeemer = Plutus.Redeemer
+
+-- | @since 2.0.0
+deriving via
+  (DerivePConstantViaBuiltin Plutus.Redeemer PRedeemer PData)
+  instance
+    PConstantDecl Plutus.Redeemer
+
+-- | @since 2.0.0
+newtype PDatumHash (s :: S) = PDatumHash (Term s PByteString)
+  deriving stock
+    ( -- | @since 2.0.0
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since 2.0.0
+      PlutusType
+    , -- | @since 2.0.0
+      PIsData
+    , -- | @since 2.0.0
+      PEq
+    , -- | @since 2.0.0
+      PPartialOrd
+    , -- | @since 2.0.0
+      POrd
+    , -- | @since 2.0.0
+      PShow
+    )
+
+-- | @since 2.0.0
+instance DerivePlutusType PDatumHash where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since 2.0.0
+instance PUnsafeLiftDecl PDatumHash where
+  type PLifted PDatumHash = Plutus.DatumHash
+
+-- | @since 2.0.0
+deriving via
+  (DerivePConstantViaBuiltin Plutus.DatumHash PDatumHash PByteString)
+  instance
+    PConstantDecl Plutus.DatumHash
+
+-- | @since 2.0.0
+newtype PRedeemerHash (s :: S) = PRedeemerHash (Term s PByteString)
+
+{- | Hash a script, appending the Plutus V2 prefix.
+
+@since 2.0.0
+-}
+scriptHash :: Script -> Plutus.ScriptHash
```

Comment:
Need opinion: Do we want to keep functions near their relevant types(e.g. hash functions near hash API data types) or do we want to keep these helper function in its own function away from all data type definitions?

Personally, I'd prefer latter; it's easier to find when it's like that

Reply:
If a function is publically-exported, I like to keep it near its type(s), but this is sometimes hard to make good decisions around. For 'hidden helpers' which don't get exported, I agree, they should be grouped together.

Reply:
If function's type includes type that is defined, let's keep it near the type. If it doesn't have the exact type, let's keep it at the bottom. This means we move functions that is exported but not directly related to any of defined type to the bottom